tokio-util/src/sync/reusable_box.rs

```diff
@@ -6,26 +6,23 @@ use std::ptr::{self, NonNull};
 use std::task::{Context, Poll};
 use std::{fmt, panic};
 
-/// A reusable `Pin<Box<dyn Future<Output = T> + Send>>`.
+/// A reusable `Pin<Box<dyn Future<Output = T> + Send + 'a>>`.
 ///
 /// This type lets you replace the future stored in the box without
 /// reallocating when the size and alignment permits this.
-pub struct ReusableBoxFuture<T> {
-    boxed: NonNull<dyn Future<Output = T> + Send>,
+pub struct ReusableBoxFuture<'a, T> {
+    boxed: NonNull<dyn Future<Output = T> + Send + 'a>,
 }
 
-impl<T> ReusableBoxFuture<T> {
+impl<'a, T> ReusableBoxFuture<'a, T> {
     /// Create a new `ReusableBoxFuture<T>` containing the provided future.
     pub fn new<F>(future: F) -> Self
     where
-        F: Future<Output = T> + Send + 'static,
+        F: Future<Output = T> + Send + 'a,
     {
-        let boxed: Box<dyn Future<Output = T> + Send> = Box::new(future);
+        let boxed: Box<dyn Future<Output = T> + Send + 'a> = Box::new(future);
 
-        let boxed = Box::into_raw(boxed);
-
-        // SAFETY: Box::into_raw does not return null pointers.
-        let boxed = unsafe { NonNull::new_unchecked(boxed) };
+        let boxed = NonNull::from(Box::leak(boxed));
```

Comment:
This is a nice change which should be incorporated even if the overall PR isn't accepted!

Reply:
I don't really like going through references though. The unsafety rules get a lot more complicated when you do.

Reply:
@Darksonn `Box::into_raw` actually uses `Box::leak` internally, and `NonNull::from(Box::leak(boxed))` was the reason that `Box::into_non_null` was decided to not be added, so I am pretty sure that this would be fine.

Reply:
Fair enough.